### PR TITLE
Formatic integration into instant

### PR DESCRIPTION
**Pull request by dave4506**
## Description

New PR rebased from latest development with Fortmatic support added in #2227 

**TODO**

- [x] Test changes work for v3
- [x] Put on dogfood
- [x] Any final changes required.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Comment by chriskalani** [comment]
The button states get all jank when you pop open Fortmatic, x out, and then tap the metamask or CB wallet button. Also busted on mobile:

IMG_0134

**Comment by Agupane** [comment]
Hi @chriskalani, thanks for the feedback. I'm not able to test the mobile version, how I can reproduce this locally?, seems you are using an aws instance which I cannot access.

**Comment by chriskalani** [comment]
@dave4506 were you going to take care of these bugs?

**Comment by dave4506** [comment]
I have fixed the mobile bug with my latest commit @Agupane, the other issue is harder to fix and to reproduce it: 
1. open fortmatic
2. close the wallet modal
3. click on metamask

clicking on metamask will still prompt fortmatic.

From my investigation it relates to the issue with how our heart beat function works; it is checking for the injected web3 state, and that is making instant misinterpret the action of using metamask. 

Hope this helps @Agupane

**Comment by Agupane** [comment]
Will check it and see what I can do, thanks @dave4506 !

**Comment by Agupane** [comment]
Hi @dave4506, I tried to see what I can do with this bug, but I couldn't reproduce it, seems that the fortmatic provider is not working anymore, once I press on fortmatic it hangs at this state:
image

I saw that the code is quite different from the last time, it's quite difficult to find the problem, do you have any suggestions how we should continue with this pr?

Thanks!

**Comment by dave4506** [comment]
@Agupane Thats odd, try rebuilding the whole monorepo quite a bit changed from the original PR, I can't reproduce what you are seeing.

**Comment by Agupane** [comment]
@dave4506 HI Dave, I just try to clone the repo again and rebuild the proyect. Followed this steps:

```
1. Yarn
2. Yarn build
3. cd /packages/instant && yarn build
```

Got this:

```
ERROR in ./src/index.umd.ts
Module not found: Error: Can't resolve '@0x/asset-swapper' in '/home/apane/Documents/Repositorios/test/0x-monorepo/packages/instant/src'
 @ ./src/index.umd.ts 50:22-50

ERROR in ./src/components/buy_button.tsx
Module not found: Error: Can't resolve '@0x/asset-swapper' in '/home/apane/Documents/Repositorios/test/0x-monorepo/packages/instant/src/components'
 @ ./src/components/buy_button.tsx 51:22-50
 @ ./src/components/buy_order_state_buttons.tsx
 @ ./src/containers/selected_asset_buy_order_state_buttons.ts
 @ ./src/components/zero_ex_instant_container.tsx
 @ ./src/components/zero_ex_instant.tsx
 @ ./src/index.ts
 @ ./src/index.umd.ts

ERROR in ./src/util/asset.ts
Module not found: Error: Can't resolve '@0x/asset-swapper' in '/home/apane/Documents/Repositorios/test/0x-monorepo/packages/instant/src/util'
 @ ./src/util/asset.ts 3:22-50
 @ ./src/components/zero_ex_instant_provider.tsx
 @ ./src/components/zero_ex_instant.tsx
 @ ./src/index.ts
 @ ./src/index.umd.ts

ERROR in ./src/util/asset_swapper_factory.ts
Module not found: Error: Can't resolve '@0x/asset-swapper' in '/home/apane/Documents/Repositorios/test/0x-monorepo/packages/instant/src/util'
 @ ./src/util/asset_swapper_factory.ts 3:22-50
 @ ./src/util/provider_state_factory.ts
 @ ./src/components/zero_ex_instant_provider.tsx
 @ ./src/components/zero_ex_instant.tsx
 @ ./src/index.ts
 @ ./src/index.umd.ts

ERROR in [at-loader] ./src/types.ts:1:58 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/redux/actions.ts:1:36 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/redux/reducer.ts:1:36 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/util/asset.ts:1:66 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/util/asset.ts:117:23 
    TS2339: Property 'amountAvailableToFill' does not exist on type 'Error'.

ERROR in [at-loader] ./src/util/asset.ts:121:58 
    TS2339: Property 'amountAvailableToFill' does not exist on type 'Error'.

ERROR in [at-loader] ./src/util/asset.ts:122:33 
    TS2339: Property 'amountAvailableToFill' does not exist on type 'Error'.

ERROR in [at-loader] ./src/util/analytics.ts:1:36 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/util/swap_quote_updater.ts:1:48 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/redux/async_data.ts:39:17 
    TS2345: Argument of type '{}[]' is not assignable to parameter of type 'string[]'.
  Type '{}' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/order_details.tsx:1:31 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/components/order_details.tsx:161:51 
    TS2339: Property 'totalTakerAssetAmount' does not exist on type 'OCType<any>'.
  Property 'totalTakerAssetAmount' does not exist on type 'IDataAccessor<any>'.

ERROR in [at-loader] ./src/containers/latest_buy_quote_order_details.ts:19:46 
    TS2339: Property 'worstCaseQuoteInfo' does not exist on type 'OCType<any>'.
  Property 'worstCaseQuoteInfo' does not exist on type 'IDataAccessor<any>'.

ERROR in [at-loader] ./src/components/buy_button.tsx:8:8 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/components/buy_button.tsx:139:47 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_button.tsx:141:54 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_button.tsx:142:60 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_button.tsx:145:55 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_button.tsx:146:52 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_button.tsx:151:50 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_button.tsx:152:44 
    TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

ERROR in [at-loader] ./src/components/buy_order_state_buttons.tsx:1:91 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/containers/selected_asset_buy_order_state_buttons.ts:1:91 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/containers/selected_erc20_asset_amount_input.ts:1:28 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/containers/selected_asset_instant_heading.ts:28:55 
    TS2339: Property 'worstCaseQuoteInfo' does not exist on type 'OCType<any>'.
  Property 'worstCaseQuoteInfo' does not exist on type 'IDataAccessor<any>'.

ERROR in [at-loader] ./src/util/asset_swapper_factory.ts:1:63 
    TS2307: Cannot find module '@0x/asset-swapper'.

ERROR in [at-loader] ./src/index.umd.ts:1:39 
    TS2307: Cannot find module '@0x/asset-swapper'.
error Command failed with exit code 2.
```

**Comment by dave4506** [comment]
You will need to run the following to be safe and get a full rebuild:
```
$ yarn run remove_node_modules
```
```
$ yarn
```
```
$ yarn run rebuild
```

**Comment by dave4506** [comment]
I just pushed a update that largely fixes the Fortmatic integration; the key issue we are seeing though is how Fortmatic provider's `getAvailableAddressesAsync` function works.

Instant expects that if the user has not unlocked a provider for `getAvailableAddressesAsync` to return `[]` and not bring up any modal. Only when `enable()` is called that we are concerned with a modal. 

We are seeing a bug where the Fortmatic prompt comes up twice (as in I close it, and it pops up again). That is mainly due to `getAvailableAddressesAsync`  and `enable` getting called when Fortmatic has yet to be enabled. Instant attempts to open up the modal twice.

Is there any insight on your part @Agupane as to how we can fix that or augment Fortmatic behavior?

File: packages/instant/src/constants.ts [hunk 1]
```diff
@@ -91,5 +91,7 @@ export const PROVIDER_TYPE_TO_NAME: { [key in ProviderType]: string } = {
     [ProviderType.Parity]: 'Parity',
     [ProviderType.TrustWallet]: 'Trust Wallet',
     [ProviderType.Opera]: 'Opera Wallet',
+    [ProviderType.Fortmatic]: 'Fortmatic',
     [ProviderType.Fallback]: 'Fallback',
 };
+export const FORTMATIC_API_KEY = process.env.INSTANT_FORTMATIC_API_KEY || 'pk_live_43E27F282B16AA8C';
```
Comment (kimpers) [inline packages/instant/src/constants.ts]: Should this have a hardcoded fallback?

**Pull request merged**